### PR TITLE
Fix: Build only the specified artifact library when multiple types are available

### LINKED ISSUE #12109: artifact-dependencies seems to build both cdylib and staticlib when only one is specified
### Problem

When an artifact dependency specifies `artifact='staticlib'`, and the dependency declares its crate type as `["staticlib", "cdylib"]`, cargo will build both the staticlib and cdylib separately. I would expect it to only build the staticlib.

### Steps

Create a project with an artifact dependency:

```toml
[package]
name = "foo"
version = "0.1.0"
edition = "2021"

[dependencies]
bar = {path = "bar", artifact = "staticlib"}
```

and in the bar dependency, specify it with multiple crate types:

```toml
[package]
name = "bar"
version = "0.1.0"
edition = "2021"

[lib]
crate-type = ["staticlib", "cdylib"]
```

### Possible Solution(s)

_No response_

### Notes

_No response_

### Version

```text
cargo 1.71.0-nightly (569b648b5 2023-05-05)
release: 1.71.0-nightly
commit-hash: 569b648b5831ae8a515e90c80843a5287c3304ef
commit-date: 2023-05-05
host: aarch64-apple-darwin
libgit2: 1.6.3 (sys:0.17.0 vendored)
libcurl: 7.87.0 (sys:0.4.61+curl-8.0.1 system ssl:(SecureTransport) LibreSSL/3.3.6)
ssl: OpenSSL 1.1.1t  7 Feb 2023
os: Mac OS 13.3.1 [64-bit]
```

### PR DESCRIPTION
**Pull request by skogseth**
### What does this PR try to resolve?
Fixes #12109. 

#### TL;DR:
A crate `bar` exposes it's library as both a `staticlib` and a `cdylib`. A second crate `foo` specifies an artifact dependency of type `staticlib` on `bar`, meaning cargo should build `bar` as a `staticlib` and expose certain environment variables (e.g. `CARGO_STATICLIB_FILE_BAR`). However, due to a bug, cargo ends up building (and exposing the associated environment variables) for both the `staticlib` and `cdylib`. 

The same happens if `foo` specifies an artifact dependency of type `cdylib`; both artifact types are built.

### How should we test and review this PR?
The first commit introduces a test which reproduces the issue, the second commit introduces the fix. This setup was recommended by @ehuss.

### Additional information
Artifact dependencies: https://rust-lang.github.io/rfcs/3028-cargo-binary-dependencies.html

#### TL;DR
If a crate `foo` requests an artifact dependency of kind <artifact_kind> from a crate `bar` then the following happens:

- `bar` is built with crate-type <artifact_kind> and a directory is created at `target/<profile>/deps/artifact/bar-<build_hash_or_something>/<artifact_kind>`. The binary artifact is placed in this directory.
- Cargo exposes certain environment variables, the most important for this PR is `CARGO_<artifact_kind>_FILE_BAR`, which points to the binary artifact that was specified. This environment variable is available at compile time for normal dependencies and at runtime for build-dependencies.

If multiple artifact-kinds are requested cargo will create a unit for each, and so they will all be built separately.

### CODE REVIEW

File: tests/testsuite/artifact_dep.rs [hunk 1]
```diff
@@ -3189,3 +3189,77 @@ fn check_transitive_artifact_dependency_with_different_target() {
...
+            .build()
+    };
+
+    let cdylib = create_project("cdylib");
+    cdylib
+        .cargo("build -Z bindeps")
+        .masquerade_as_nightly_cargo(&["bindeps"])
+        .run();
+    match_exact(
+        "cdylib present: true\nstaticlib present: false",
```
Comment (weihanglo) [inline tests/testsuite/artifact_dep.rs:3244]: > Just make sure it has a test, preferably as the first commit and then include the fix in the second commit to show how the test changes.

So this should be

```suggestion
        "cdylib present: true\nstaticlib present: true",
```

in the first commit. And the next commit would fix both the behavior and this test. This setup is better because:

* Every commit passes CI pipeline.
* Changes of the test shows we really fixes the behavior.

File: tests/testsuite/artifact_dep.rs [hunk 2]
```diff
@@ -3189,3 +3189,77 @@ fn check_transitive_artifact_dependency_with_different_target() {
...
+        None,
+    )
+    .unwrap();
+
+    let staticlib = create_project("staticlib");
+    staticlib
+        .cargo("build -Z bindeps")
+        .masquerade_as_nightly_cargo(&["bindeps"])
+        .run();
+    match_exact(
```
Comment (weihanglo) [inline tests/testsuite/artifact_dep.rs:3257]: Not sure why this test module chose to use match_exact but not  `cargo_test_support::compare::assert_match_exact`. Anyway, I'll find a time clean this up. For now we just follow what has been written :)

File: tests/testsuite/artifact_dep.rs [hunk 3]
```diff
@@ -3189,3 +3189,77 @@ fn check_transitive_artifact_dependency_with_different_target() {
...
+        "cdylib present: false\nstaticlib present: true",
```
Comment (weihanglo) [inline tests/testsuite/artifact_dep.rs:3258]: ditto

```suggestion
        "cdylib present: true\nstaticlib present: true",
```

File: src/cargo/core/compiler/unit_dependencies.rs [hunk 4]
```diff
@@ -555,17 +555,20 @@ fn artifact_targets_to_unit_deps(
...
                 match target.kind() {
                     TargetKind::Lib(kinds) => Box::new(
                         kinds
                             .iter()
-                            .filter(|tk| matches!(tk, CrateType::Cdylib | CrateType::Staticlib))
+                            .filter(|tk| match tk {
+                                CrateType::Cdylib => *artifact_kind == ArtifactKind::Cdylib,
+                                CrateType::Staticlib => *artifact_kind == ArtifactKind::Staticlib,
+                                _ => false,
+                            })
```
Comment (weihanglo) [inline src/cargo/core/compiler/unit_dependencies.rs:571]: What about this? 

(just personal taste, not too important)

```suggestion
                            .filter(move |tk| match (tk, artifact_kind) {
                                (CrateType::Cdylib, ArtifactKind::Cdylib) => true,
                                (CrateType::Staticlib, ArtifactKind::Staticlib) => true,
                                _ => false,
                            })
```

**Review by weihanglo — COMMENTED** [review]
Overall looks good! I added some minor suggestions.

Thanks for your contribution!

File: tests/testsuite/artifact_dep.rs (hunk 1 above)
Comment (skogseth) [reply tests/testsuite/artifact_dep.rs:3244]: Ahh, that makes a lot of sense!

File: src/cargo/core/compiler/unit_dependencies.rs (hunk 4 above)
Comment (skogseth) [reply src/cargo/core/compiler/unit_dependencies.rs:571]: Ooohh, I like it

**Review by weihanglo — APPROVED** [review]

**Comment by weihanglo** [comment]
Thank you for the patch!

@bors r+
